File: solutions/javascript/pig-latin/1/pig-latin.js
export const translate = (phrase) => {
  return phrase.split(' ').map(word => {
    if (/^([aeiou]|xr|yt)/.test(word)) {
      return word + 'ay';
    }

    const quMatch = word.match(/^([^aeiou]*qu)(.*)/);
    if (quMatch) {
      return quMatch[2] + quMatch[1] + 'ay';
    }

    const yMatch = word.match(/^([^aeiou]+)(y.*)/);
    if (yMatch) {
      return yMatch[2] + yMatch[1] + 'ay';
    }

    const consMatch = word.match(/^([^aeiou]+)(.*)/);
    if (consMatch) {
      return consMatch[2] + consMatch[1] + 'ay';
    }

    return word;
  }).join(' ');
};

File: solutions/kotlin/pig-latin/1/src/main/kotlin/PigLatin.kt
object PigLatin {
    fun translate(phrase: String): String {
        return phrase.split(" ").joinToString(" ") { translateWord(it) }
    }

    private fun translateWord(word: String): String {
        val vowels = setOf('a', 'e', 'i', 'o', 'u')

        // Правило 1: Начинается с гласной, "xr" или "yt"
        if (word[0] in vowels || word.startsWith("xr") || word.startsWith("yt")) {
            return word + "ay"
        }

        // Правило 3: Начинается с согласных, за которыми следует "qu"
        val quIndex = word.indexOf("qu")
        if (quIndex != -1 && word.substring(0, quIndex).all { it !in vowels }) {
            val prefix = word.substring(0, quIndex + 2)
            val suffix = word.substring(quIndex + 2)
            return suffix + prefix + "ay"
        }

        // Правило 4: Начинается с согласных, за которыми следует "y"
        // 'y' не должна быть самой первой буквой (иначе это обработается как согласная в Правиле 2)
        val yIndex = word.indexOf('y')
        if (yIndex > 0 && word.substring(0, yIndex).all { it !in vowels }) {
            val prefix = word.substring(0, yIndex)
            val suffix = word.substring(yIndex)
            return suffix + prefix + "ay"
        }

        // Правило 2: Начинается с одной или нескольких согласных
        var consonantsCount = 0
        while (consonantsCount < word.length && word[consonantsCount] !in vowels) {
            consonantsCount++
        }
        if (consonantsCount > 0) {
            val prefix = word.substring(0, consonantsCount)
            val suffix = word.substring(consonantsCount)
            return suffix + prefix + "ay"
        }

        return word
    }
}
